report/report.tex

```diff
@@ -26,24 +27,180 @@
+
+\subsubsection{Procedure Declarations}
+Every Hanus procedure declaration is compiled to a Haskell functions, consisting of a single do-statement, that performs a computation that is equivalent to that intended by the Hanus procedure. Every Hanus statement in a procedure is converted to one, or sometimes more than one statement in the Haskell do block. Since a procedure can be called to run it regularly and uncalled to run it in reverse, every Hanus procedure $p$ is compiled to not one but two Haskell functions: a function $p$ that does exactly what Hanus procedure $p$ should do, and a function $p'$ that is the inverse of $p$. Name conflicts as a result of this convention are impossible since the Hanus parser does not accept the $'$ character as part of a procedure identifier.
+
+\subsubsection{Global variables}
+Hanus has two types of variables. Global variables are declared at the top level with a name and type. The programmer cannot specify a default value for the variable, but instead the variable will automatically get the default value of the specified type. Since global variables can be updated from anywhere in a Hanus program, Hanus imposes the restriction on the programmer that global variables cannot be passed to procedures. Furthermore, it would seem to be a sensible choice to substitute \emph{IORefs} for global variables since \emph{IORefs} are a Haskell solution to variables. However, the problem would then be that the entire Hanus AST and possible Haskell sub-AST's (since any Haskell expression can be written on the right hand side of a Hanus expression) would have to be traversed, replacing all variable declarations and expressions with \emph{readIORef} and \emph{writeIORef} operations, which would have been quite a challenge. Instead, global variables are simply declared as lets. If they need to be updated, they are just redeclared. Note that something like \inlinecode{a += 1} is not compiled to \inlinecode{let a = a + 1}, because this causes an endless loop due to \inlinecode{a} being on the left and right hand side. Instead, it is compiled to \inlinecode{let temp = a; let a = temp + 1}. To handle all difficult cases where there are function calls or nested statement blocks that (possibly) shadow the let declaration, but only in the nested scope, a simple call-by-copy-restore mechanism is used: the Haskell counterpart of any Hanus procedure takes all global variables as inputs, and the last statement in the do-block of the Haskell function returns all global variables in a single tuple. The callee then redeclares all global variables after the procedure call. An example:
+                        
+\begin{minted}{haskell}
+a :: Int;
+b :: Int;
+procedure x() {
+    call y;
+}
+procedure y() {
+    a += 1;
+    b += 1;
+}
+\end{minted}
+
+\noindent This procedure is compiled to something like the following:
+
+\begin{minted}{haskell}
```

Comment:
I used the \code command from the poster in for my examples, maybe we should chose one or the other, but only one to remain consistent.

Reply:
There are already code examples from Ivo in the master branch that use minted. I suggest that we use minted for the smaller examles (including the ones in this PR) and use your fancy \code command for the larger examples that you wrote; it's a bit overkill to add line numbers and everyting to relatively small examples in my opinion.

Reply:
I'm fine with that solution